### ISSUE TITLE
Setup Continuous Integration (CI): Automated testing for pushes and pull requests 

### ISSUE THREAD

**Issue by niketagrawal** [tudat-team/tudat#199]
### Current behavior
Tests are not triggered on push or pull request to develop and master. Tests are only run when code changes are made to the tudat feedstock repository, i.e., when a conda package is published.

### Desired behavior
Every pull request to develop or master and push to feature branches should automatically trigger build, run tests, and indicate success/failure status.

Implementing Continuous Integration (CI) by setting up automated testing in the tudat repository will allow the contributors and the reviewers to see whether the new code changes preserve the expected functionality of the software. 	

### Implementation

Speed up CI workflow across subsequent runs by adding the following to the workflow:
- [x]  Cache previous compilation results using ccache
- [x] Cache conda environment

**Comment by LeonardoPohl** [comment]
A great idea! 

I was wondering if such a system might also be useful for tudat bundle, automatically building the kernel from source. I am aware that this proposal is more meant as a check for the code functionality rather than releases, but I think it could be extended if there is a wish/need for this.

I think this would require the respective master branches of tudat and tudatpy to be compilable, so this would still issue would still be quite useful and also for tudatpy. If both repositories are basically 'required' to be compilable by a CI like it is proposed here, the creation of releases will be less problematic.

**Comment by DominicDirkx** [comment]
That's a great point, and I think setting up a similar CI for tudatpy would be great! 

I'm not sure about CI for tudat-bundle, since this is a repo that is (for now) exclusively used as a tool for developers to allow them to develop tudat and tudatpy in parallel. The release system we currently have for conda packages is run on Azure, where the tudatpy build uses the latest tudat conda package. If this latest conda package is not compatible with the tudatpy code being run, compilation fails. I'd imagine something similar being used for tudatpy CI.

But, this does raise the issue of how to test a tudat and tudatpy pull request that add the same piece of functionality to the two repos. For this, a CI system on tudat-bundle may indeed be useful, but there would need to be some system in place linking a tudat and tudatpy pull request to one another. This could be a nice feature to have, any ideas on this?

**Comment by LeonardoPohl** [comment]
As for the first point. To avoid duplicate code we can setup a third repository which contains the [reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows), if the process is similar in both repositories. How much sense this makes in these early stages is a different question, but maybe something to keep in mind when developing a solution which requires both tudatpy and tudat repository to be at the same stage of 'compilability' (Im not sure if thats a word).

As for the second point, I myself have struggled because of an issue stemming from this when I first tried to build the system and I think the project would benefit from that. Is there a way of getting an output from the current system in azure to avoid duplicate builds?

If the CI is setup in tudat-bundle I believe the two repositories can be pulled within the workflow, using action/checkout. If I remember correctly though, a security token needs to be setup in the repositories, the last time I did this it took a bit to get working properly but its doable. And from there the build script can be used.
